### PR TITLE
Restart at eor

### LINKED ISSUE #17: time:units set inconsistently in the nuopc interface
**Describe the bug**
In the netcdf output, when using a restart file, then time:units attribute is set in a way that is inconsistent with the time variable.

For an initial run of one day long, in the netcdf output:
```
time:units = "seconds since 1900-01-01 00:00:00" ; 
time = 86400 ;
```
Which is correct

If another day is run, the time variable in netcdf output is:
```
time:units = "seconds since 1900-01-02 00:00:00" ;
time = 172800 ;
```
This is inconsistent, as both the time and the time origin have stepped forward one day.

Interestingly the file name has the correct date (i.e. the end of the time period):
`GMOM_JRA_WD.ww3.hi.1900-01-03-00000.nc`

**To Reproduce**
Start the model from a restart file, with netcdf output turned on.

**Expected behavior**
Time:units should be set from the model start time, not the restart time.

**Additional context**

The `time_origin` is set based on the 'Current Time' before the model is time stepped:

https://github.com/ESCOMP/WW3/blob/97740a1abb36b7ba6e52a7d6f7a37dd06f915633/model/src/wav_comp_nuopc.F90#L613
```
   call ESMF_ClockGet( clock, currTime=esmfTime, rc=rc )
      if (ChkErr(rc,__LINE__,u_FILE_u)) return
    endif
    ! Determine time attributes for history output
    call ESMF_TimeGet( esmfTime, timeString=time_origin, calendar=calendar, rc=rc )
    if (ChkErr(rc,__LINE__,u_FILE_u)) return
    time_origin = 'seconds since '//time_origin(1:10)//' '//time_origin(12:19)
```

But then the 'elapsed_secs' is set from the 'Start Time'

https://github.com/ESCOMP/WW3/blob/97740a1abb36b7ba6e52a7d6f7a37dd06f915633/model/src/wav_comp_nuopc.F90#L1062

```
elapsedTime = nextTime - startTime
    call ESMF_TimeIntervalGet(elapsedTime, s_i8=elapsed_secs,rc=rc)
```

I think setting time_origin based on Start Time would make sense.

### PR DESCRIPTION
**Pull request by jedwards4b**
# Pull Request Summary
Adds support for cesm write_restart_at_endofrun

## Description
Denise Worthen may want to review.  The stop and restart alarm initialization order needs to be reversed this is in support of setting a restart alarm based on the stop time.   A new function get_minimum_time is needed in wav_shr_mod to correct timestamp errors when using _OPTION=nsteps.  

### Commit Message
The CESM options REST_OPTION=nsteps, STOP_OPTION=nsteps and HIST_OPTION=nsteps were not properly set by WW3.  The component model needs to know the minimum model coupling time in order to correctly set this option.

### Check list  

- [X] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [ ] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ ] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [ ] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?
* Please indicate the expected changes in the regression test output, (Note the [list](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-develop#4-look-at-results) of known non-identical tests.)
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):

### CODE REVIEW

File: model/src/wav_shr_mod.F90 [hunk 1]
```diff
@@ -990,10 +1001,22 @@ subroutine alarmInit( clock, alarm, option, &
...
+      if (ChkErr(rc,__LINE__,u_FILE_u)) return
       AlarmInterval = AlarmInterval * opt_n
+      ! timestepinterval*0 is 0 of kind ESMF_TimeStepInterval
+      if (mod(AlarmInterval, TimestepInterval) /= (timestepinterval*0)) then
+         call ESMF_LogWrite(subname//'illegal Alarm setting for '//trim(alarmname), ESMF_LOGMSG_ERROR)
+         rc = ESMF_FAILURE
+         return
+      endif
       update_nextalarm  = .true.
+!      call ESMF_ClockGet(clock, TimeStep=AlarmInterval, rc=rc)
```
Comment (mvertens) [inline model/src/wav_shr_mod.F90]: Why keep the comment?

File: model/src/wav_shr_mod.F90 [hunk 2]
```diff
@@ -1343,6 +1374,51 @@ subroutine ymd2date_long(year,month,day,date)
...
+  integer function get_minimum_timestep(gcomp, rc)
+    ! Get the minimum timestep interval in seconds based on the nuopc.config variables *_cpl_dt,
+    ! if none of these variables are defined this routine will throw an error
+    type(ESMF_GridComp), intent(in) :: gcomp
+    integer, intent(out)            :: rc
+
+    character(len=CS) :: cvalue
+    integer                 :: comp_dt             ! coupling interval of component
+    integer, parameter :: ncomps = 8
+    character(len=3),parameter :: compname(ncomps) = (/"atm", "lnd", "ice", "ocn","glc","rof", "wav", "esp"/)
```
Comment (mvertens) [inline model/src/wav_shr_mod.F90]: Why is this routine in wav share code and not in the general share code? I thought there was actually a routine now in share that could deal with this.

**Review by mvertens — CHANGES_REQUESTED** [review]
I think this looks okay - except for asking why the get_min_timestep is in wave specfic code when it could be in the general share code?

**Review by alperaltuntas — APPROVED** [review]
Other than @mvertens's comments, lgtm.

**Comment by jedwards4b** [comment]
I have addressed her comment, and tested the change.   I will push it back now.

File: model/src/wav_shr_mod.F90 (hunk 2 above)
Comment (jedwards4b) [reply model/src/wav_shr_mod.F90]: removed

File: model/src/wav_shr_mod.F90 (hunk 1 above)
Comment (jedwards4b) [reply model/src/wav_shr_mod.F90]: removed

**Review by mvertens — APPROVED** [review]

**Comment by alperaltuntas** [comment]
@jedwards4b Is this PR ready to be merged?  I think, moving forward, we should use the assignee feature to specify who will merge the PR, whether it's author themselves or someone else, in addition to assigning reviewers.

**Comment by jedwards4b** [comment]
Yes, please merge.

**Pull request merged**